File: frontend/utils/bitId.ts
export const bitIdToIds = (stringBitId: string): number[] => {
  let bitId = BigInt(stringBitId)
  const ids: number[] = []
  let shift = 0

  while (bitId > 0n) {
    if (bitId & 1n) {
      ids.push(shift + 1)
    }

    bitId >>= 1n
    shift += 1
  }

  return ids
}

const idToBitId = (id: number): bigint => 1n << BigInt(id - 1)

export const idsToBitId = (ids: number[]): string => String(ids.reduce((current, id) => current | idToBitId(id), 0n))

File: lib/diver_down/web/bit_id.rb
# frozen_string_literal: true

module DiverDown
  class Web
    # Use bitflag for a lightweight representation of a list of bound definitions on a URL.
    # Each definition is related with a unique bitflag, and when merging, the bitflag is calculated with `OR` to represent the merged definitions in a lightweight and fast.
    module BitId
      class << self
        # @param ids [Array<Integer>]
        # @return [Integer]
        def ids_to_bit_id(ids)
          ids.inject(0) { _1 | id_to_bit_id(_2) }
        end

        # @param id [Integer]
        # @return [Array<Integer>]
        def bit_id_to_ids(bit_id)
          ids = []
          shift = 0
          while bit_id.positive?
            if bit_id & 1
              ids.push(shift + 1)
            end

            bit_id >>= 1
            shift += 1
          end
          ids
        end

        private

        # @param id [Integer]
        # @return [Integer]
        def id_to_bit_id(id)
          1 << (id - 1)
        end
      end
    end
  end
end
